language/gpt3/text_generation_server.py

```diff
@@ -0,0 +1,206 @@
+                        _,
+                        _,
+                    ) = generate_tokens_probs_and_return_on_first_stage(
+                        self.model,
+                        input_ids_tensor,
+                        input_length_tensor,
+                        top_k=self.gen_kwargs.get("top_k", 4),
+                        # top_p = self.gen_kwargs.get("top_p", 0.9),
+                        temperature=self.gen_kwargs.get("temperature", 0.0),
+                    )
+                    output_batch_truncated = []
+                    for data, source_len in zip(output_tokens, input_length_tensor):
+                        output_batch_truncated.append(
+                            data[source_len:].cpu().numpy().tolist()
+                        )
+                    if self.log:
+                        print("end time: ", datetime.datetime.now())
+                    return jsonify({"output": output_batch_truncated})
+
+            except ValueError as ve:
+                return ve.args[0]
+
+
+class MegatronServer(object):
+    def __init__(self, model, gen_kwargs):
+        self.app = Flask(__name__, static_url_path="")
+        api = Api(self.app)
+        api.add_resource(
+            MegatronGenerate, "/api", resource_class_args=[model, gen_kwargs]
+        )
+
+    def run(self, url):
+        self.app.run(url, threaded=True, debug=False)
+
+
+def model_provider(pre_process=True, post_process=True):
+    """Build the model."""
+
+    print_rank_0("building GPT model ...")
+    model = GPTModel(
+        num_tokentypes=0,
+        parallel_output=False,
+        pre_process=pre_process,
+        post_process=post_process,
+    )
+
+    return model
+
+
+if __name__ == "__main__":
+    initialize_megatron(
+        args_defaults={
+            "tokenizer_type": "GPT2BPETokenizer",
+            "no_load_rng": True,
+            "no_load_optim": True,
+        }
+    )
+
+    args = get_args()
+    gen_kwargs = {
+        "early_stopping": True,
+        "max_new_tokens": 128,
+        "min_new_tokens": 30,
+        "top_k": 4,
+        "top_p": 0.9,
+        "temperature": 0.0,
+    }
+    if args.num_layers_per_virtual_pipeline_stage is not None:
+        print("Interleaved pipeline schedule is not yet supported for text generation.")
+        exit()
+    # Set up model and load checkpoint
+    model = get_model(model_provider, wrap_with_ddp=False)
+
+    if args.load is not None:
+        _ = load_checkpoint(model, None, None)
+
+    assert len(model) == 1, "Above condition should have caught this"
+    model = model[0]
+    if mpu.is_pipeline_first_stage() and mpu.get_tensor_model_parallel_rank() == 0:
+        server = MegatronServer(model, gen_kwargs)
+        server.run("0.0.0.0")
+
+    while True:
+        choice = torch.cuda.LongTensor(1)
+        input_length_tensor = torch.cuda.LongTensor(1)
+        torch.distributed.broadcast(choice, 0)
+        if choice[0].item() == 0:
```

Comment:
This flow is a bit unclear. Could you add some comments?

What happens if ``choice[0].item()==0``, what if ``choice[0].item()==1``? 
When ``choice[0].item()==0``?

Reply:
This decides which decoding method to use. Beam search was just implemented